lib/model/message.dart

```diff
@@ -66,6 +234,219 @@ mixin MessageStore {
+  void _updateOutboxMessage(int localMessageId, {
+    required OutboxMessageState newState,
+  }) {
+    assert(!_disposed);
+    final outboxMessage = outboxMessages[localMessageId];
+    if (outboxMessage == null) {
+      throw StateError(
+        'Removing unknown outbox message with localMessageId: $localMessageId');
+    }
+    final oldState = outboxMessage.state;
+    // See [OutboxMessageState] for valid state transitions.
+    final isStateTransitionValid = switch (newState) {
+      OutboxMessageState.hidden => false,
+      OutboxMessageState.waiting =>
+        oldState == OutboxMessageState.hidden,
+      OutboxMessageState.waitPeriodExpired =>
+        oldState == OutboxMessageState.waiting,
+      OutboxMessageState.failed =>
+        oldState == OutboxMessageState.hidden
+        || oldState == OutboxMessageState.waiting
+        || oldState == OutboxMessageState.waitPeriodExpired,
+    };
+    if (!isStateTransitionValid) {
+      throw StateError('Unexpected state transition: $oldState -> $newState');
+    }
+
+    outboxMessage._state = newState;
+    for (final view in _messageListViews) {
+      if (oldState == OutboxMessageState.hidden) {
+        view.addOutboxMessage(outboxMessage);
+      } else {
+        view.notifyListenersIfOutboxMessagePresent(localMessageId);
+      }
+    }
+  }
+
+  /// Send a message and create an entry of [OutboxMessage].
+  Future<void> _outboxSendMessage({
+    required MessageDestination destination,
+    required String content,
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    assert(!_disposed);
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final conversation = switch (destination) {
+      StreamDestination(:final streamId, :final topic) =>
+        StreamConversation(
+          streamId,
+          _processTopicLikeServer(
+            topic, realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+      DmDestination(:final userIds) => DmConversation(allRecipientIds: userIds),
+    };
+
+    _outboxMessages[localMessageId] = OutboxMessage.fromConversation(
+      conversation,
+      localMessageId: localMessageId,
+      selfUserId: selfUserId,
+      timestamp: ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000,
+      contentMarkdown: content);
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(
+      kLocalEchoDebounceDuration,
+      () => _handleOutboxDebounce(localMessageId));
+
+    _outboxMessageWaitPeriodTimers[localMessageId] = Timer(
+      kSendMessageOfferRestoreWaitPeriod,
+      () => _handleOutboxWaitPeriodExpired(localMessageId));
+
+    try {
+      await _apiSendMessage(connection,
+        destination: destination,
+        content: content,
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
```

Comment:
This feels a bit mysterious on its own, so let's add a comment explaining why it's here:
```suggestion

      // The send request succeeded, so the message was definitely sent.
      // Cancel the timer that would have had us start presuming that the
      // send might have failed.
      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
```

Reply:
and nit: move this outside of the try/catch, to make clear the try/catch is focused on exceptions from `_apiSendMessage` (where exceptions are expected) and not potential exceptions from this line (which shouldn't be possible). Discussion here:
https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#catch-specific

Reply:
Writing out that comment also caused me to have the thought: what happens if the request succeeds and the state has *already* moved to `waitPeriodExpired`? The expiry means we've already started presuming that it probably failed, but suddenly we know for sure it actually succeeded.

If there isn't a quick change to make that would handle that, it's fine to leave a TODO comment for it instead.

Reply:
> and nit: move this outside of the try/catch, to make clear the try/catch is focused on exceptions from `_apiSendMessage` (where exceptions are expected) and not potential exceptions from this line (which shouldn't be possible). Discussion here: https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#catch-specific

Ah. This reminds me of adding a `if (_disposed)` check.

Reply:
> Writing out that comment also caused me to have the thought: what happens if the request succeeds and the state has _already_ moved to `waitPeriodExpired`? The expiry means we've already started presuming that it probably failed, but suddenly we know for sure it actually succeeded.
> 
> If there isn't a quick change to make that would handle that, it's fine to leave a TODO comment for it instead.


I think this would be a matter of implementing a new state transition:

 ```                                                                         
                             Got an [ApiRequestException].                   
          ┌──────┬────────────────────────────┬──────────► failed            
          │      │                            │              │               
          │      │      [sendMessage]         │              │               
 (create) │      │      request succeeds.     │              │               
    └► hidden   waiting ◄─────────────── waitPeriodExpired ──┴─────► (delete)
          │      ▲   │                     ▲               User restores     
          └──────┘   └─────────────────────┘               the draft.        
         Debounce     [sendMessage] request                                  
         timed out.   not finished when                                      
                      wait period timed out.                                 
                                                                             
              Event received.                                                
 (any state) ─────────────────► (delete)                                     
 ```                                                                         

I guess we can't really convey that `waitPeriodExpired` cannot be reached again without having a variant of `waiting`, but we can imply that from the dart doc of `waitPeriodExpired`:

```dart
  /// The [sendMessage] HTTP request did not finish in time and the user is
  /// invited to retry it.
  ///
  /// This state can be reached when the request has not finished
  /// [kSendMessageOfferRestoreWaitPeriod] since the outbox message's creation.
  waitPeriodExpired,
```

Reply:
Added a new commit `message: Avoid double-sends after send-message request succeeds`  addressing this. I found it to be a good way seeing how we can incrementally add features to this state machine.